core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * File cleanup strategy for snapshot expiration which determines, via an in-memory reference set,
+ * metadata and data files that are not reachable given the previous and current table states.
+ */
+class ReachableFileCleanup extends FileCleanupStrategy {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReachableFileCleanup.class);
+
+  private final TableMetadata afterExpiration;
+  private final TableMetadata beforeExpiration;
+
+  ReachableFileCleanup(
+      TableOperations ops,
+      TableMetadata beforeExpiration,
+      ExecutorService deleteExecutorService,
+      ExecutorService planExecutorService,
+      Consumer<String> deleteFunc) {
+    super(ops, deleteExecutorService, planExecutorService, deleteFunc);
+    this.beforeExpiration = beforeExpiration;
+    this.afterExpiration = ops.refresh();
+  }
+
+  @Override
+  public void cleanFiles() {
+    // Identify all of the manifest lists to retain
+    Set<String> manifestListsBeforeExpiration =
+        ReachableFileUtil.manifestListLocations(beforeExpiration);
+
+    // Identify all of the manifest lists that currently exist
+    Set<String> manifestListsAfterExpiration =
+        ReachableFileUtil.manifestListLocations(afterExpiration);
+
+    Set<String> manifestListsToDelete =
+        Sets.difference(manifestListsBeforeExpiration, manifestListsAfterExpiration);
+    Map<String, ManifestFile> reachableManifestsAfterExpiration =
+        computeManifestsForManifestLists(manifestListsAfterExpiration);
+
+    // Compute the manifests to delete based on the current manifest lists to delete and the current
+    // reachable manifests
+    List<ManifestFile> manifestsToDelete =
+        computeManifestsToDelete(manifestListsToDelete, reachableManifestsAfterExpiration);
+
+    List<ManifestFile> currentManifests =
+        Lists.newArrayList(reachableManifestsAfterExpiration.values());
+
+    Set<String> filesToDelete = findFilesToDelete(manifestsToDelete, currentManifests);
+    Tasks.foreach(filesToDelete)
+        .executeWith(deleteExecutorService)
+        .retry(3)
+        .stopRetryOn(NotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Delete failed for data file: {}", file, exc))
+        .run(deleteFunc::accept);
+
+    Set<String> manifestPathsToDelete =
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
```

Comment:
Same here. I think this should be removed.

Reply:
Done for reachable file cleanup, see my comment here https://github.com/apache/iceberg/pull/5669#discussion_r990676361